packages/server/api/src/app/workers/engine-controller.ts

```diff
@@ -80,6 +81,10 @@ export const flowEngineWorker: FastifyPluginAsyncTypebox = async (app) => {
     app.post('/notify-frontend', NotifyFrontendParams, async (request) => {
         const { type, data } = request.body
         app.io.to(request.principal.projectId).emit(type, data)
+        await stepRunProgressHandler.notifyStepFinished({
```

Comment:
I don't think it's correct place to place notifyStepFinished, It sounds like a side effect for flow run, I think we have place in `flow-run` when run is finished (If I am not wrong, we use it for issues).

In that place we should check If we have to send reply (Maybe it's real run which is more common), If we don't we should do early return, because this put huge burden on app server, since it have to get the run from the database and actual download the whole log from S3.